src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,258 @@
+     * @param ast {@code DetailAST} ast element
+     * @return the first sibling of the given node
+     */
+    private static DetailAST getFirstSibling(DetailAST ast) {
+        DetailAST result;
+        if (ast.getParent() == null) {
+            result = ast;
+            while (result.getPreviousSibling() != null) {
+                result = result.getPreviousSibling();
+            }
+        }
+        else {
+            result = ast.getParent().getFirstChild();
+        }
+        return result;
+    }
+
+    /**
+     * Returns list of nodes matching defined line and column number.
+     * @param root {@code DetailAST} root ast
+     * @param lineNumber line number
+     * @param columnNumber column number
+     * @return list of nodes matching defined line and column number
+     */
+    private static List<DetailAST> getMatchingAstElements(DetailAST root, int lineNumber,
+                                                          int columnNumber) {
+        final List<DetailAST> result = new ArrayList<>();
+        DetailAST curNode = root;
+        while (curNode != null && curNode.getLineNo() <= lineNumber) {
+            if (curNode.getLineNo() == lineNumber && curNode.getColumnNo() == columnNumber
+                    && curNode.getType() != TokenTypes.IDENT) {
+                result.add(curNode);
+            }
+            DetailAST toVisit = curNode.getFirstChild();
+            while (curNode != null && toVisit == null) {
+                toVisit = curNode.getNextSibling();
+                if (toVisit == null) {
+                    curNode = curNode.getParent();
+                }
+            }
+
+            curNode = toVisit;
+        }
+        return result;
+    }
+
+    /**
+     * Returns relative xpath query for given ast element from root.
+     * @param root {@code DetailAST} root element
+     * @param ast {@code DetailAST} ast element
+     * @return relative xpath query for given ast element from root
+     */
+    private static String getXpathQuery(DetailAST root, DetailAST ast) {
+        String result = "";
+        DetailAST cur = ast;
+        while (cur != root) {
+            String curNodeQuery = '/' + TokenUtils.getTokenName(cur.getType());
+            final DetailAST identAst = cur.findFirstToken(TokenTypes.IDENT);
+            if (identAst != null) {
+                curNodeQuery += "[@text='" + identAst.getText() + "']";
+            }
+            result = curNodeQuery + result;
+            cur = cur.getParent();
+        }
+        return result;
+    }
+
+    /**
+     * Checks if the given ast element has unique {@code TokenTypes} among siblings.
+     * @param firstSibling {@code DetailAST} first sibling
+     * @param ast {@code DetailAST} ast element
+     * @return if the given ast element has unique {@code TokenTypes} among siblings
+     */
+    private static boolean hasNoSiblingWithSameTokenType(DetailAST firstSibling, DetailAST ast) {
```

Comment:
Idea was to iterate all nodes on the same level, and find if there any node with same token type. If we start iterating from next sibling, probably we can miss previous nodes, since it is not surely that `DetailAST ast` is the first child

I am trying to generate test to prove my point